providers/implementations/kem/rsa_kem.c

```diff
@@ -329,13 +347,24 @@ static int rsasve_recover(PROV_RSA_CTX *prsactx,
         return 1;
     }
 
-    /* Step (2): check the input ciphertext 'inlen' matches the nlen */
+    /*
+     * Step (2): check the input ciphertext 'inlen' matches the nlen
+     * and that outlen is at least nlen bytes
+     */
     if (inlen != nlen) {
         ERR_raise(ERR_LIB_PROV, PROV_R_BAD_LENGTH);
         return 0;
     }
+    if (outlen == NULL || *outlen < nlen) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_OUTPUT_LENGTH);
+        return 0;
+    }
+
     /* Step (3): out = RSADP((n,d), in) */
-    return (RSA_private_decrypt(inlen, in, out, prsactx->rsa, RSA_NO_PADDING) > 0);
+    ret = RSA_private_decrypt(inlen, in, out, prsactx->rsa, RSA_NO_PADDING);
+    if (ret > 0)
```

Comment:
I think this should be `ret >= 0`, the return code 0 means the plain text was empty (at least that's what manual says about RSA_private_decrypt().

Reply:
I think thats a bit fuzzy.  RSA_private_decrypt indicates that a returned value of zero isn't a failure:
https://docs.openssl.org/3.1/man3/RSA_public_encrypt/#description
```
 A return value of 0 is not an error and means only that the plaintext was empty.
```

But we're returning a value for EVP_PKEY_decapsulate here, which says 0 is a failure.  I would interpret empty plaintext in decapsulate as an error.  Also, given the proximity to our beta release, I would prefer to avoid changing the conditions under which we return failure here until the next release

Reply:
I'd also note that the `> 0` is pre-existing behaviour. If we think that is wrong, then maybe raise an issue and discuss it properly rather than in a comment on a pull request that's addressing a smaller issue (same below)

Reply:
Sorry, you are both right. I confused return value of `RSA_private_decrypt()` with expectation of `EVP_PKEY_decapsulate()` which is `EVP_PKEY_decapsulate(3)`
```
EVP_PKEY_decapsulate_init(), EVP_PKEY_auth_decapsulate_init() and EVP_PKEY_decapsulate() return 1 for success and 0 or a negative value for failure. In particular a return value of -2 indicates the operation is not supported by the private key algorithm.
```

all looks good.